haskell/test/Test/TimedOracles.hs

```diff
@@ -0,0 +1,356 @@
+    consumer <- originateSimple "consumer" [] contractConsumer
+    call cfmm (Call @"Observe") $ mkView checkedTimes consumer
+    cumulatives <- getStorage consumer >>= \case
+      [v] -> pure v
+      _ -> failure "Expected to get exactly 1 cumulatives value"
+
+    let adjacents = groupAdjacent cumulatives
+    let diffs = uncurry (-) <$> adjacents
+    assert (length (List.group diffs) == 1) $
+      "Got irregular cumulatives growth:\n\
+      \  " +| List.group diffs |+ "\n\
+      \Cumulative values:\n" +| "\n\
+      \  " +| cumulatives |+ ""
+
+test_TimeOutOfBounds :: TestTree
+test_TimeOutOfBounds =
+  nettestScenarioCaps "Observing time out of bounds" do
+    alice <- newAddress "alice"
+    (cfmm, _) <- prepareSomeSegCFMM defaultStorage
+      { sCumulativesBuffer = initCumulativesBuffer 100
+      } alice
+
+    now <- getNow
+    consumer <- originateSimple "consumer" [] contractConsumer
+
+    expectFailedWith observeFutureTimestampErr $
+      call cfmm (Call @"Observe") $ mkView [now `timestampPlusSeconds` 1000] consumer
+
+    expectFailedWith observeOutdatedTimestampErr $
+      call cfmm (Call @"Observe") (mkView [now `timestampPlusSeconds` (-100000)] consumer)
+
+test_IncreaseObservationCount :: TestTree
+test_IncreaseObservationCount =
+  nettestScenarioOnEmulatorCaps "Increasing observation count works as expected" do
+    advanceTime (sec 1000000)
+
+    alice <- newAddress "alice"
+    (cfmm, _) <- prepareSomeSegCFMM defaultStorage alice
+
+    -- This helps to distinguish dummy and true values in the buffer
+    -- Note: this also triggers the contract to record a value in the buffer
+    withSender alice do
+      call cfmm (Call @"Set_position") $
+        setPositionParamSimple (TickIndex (-100), TickIndex 100) 100
+
+    -- Run invariants that can be checked immediately,
+    -- and return info (current storage) for performing later mass checks.
+    let
+      runInvariantsChecks :: (MonadEmulated caps base m, HasCallStack) => m Storage
+      runInvariantsChecks = do
+        s <- getFullStorage cfmm
+        checkCumulativesBufferInvariants s
+        return s
+
+    storageSnapshotInit <- do
+      CumulativesBufferRPC{..} <- sCumulativesBufferRPC <$> getStorage cfmm
+      cbReservedLengthRPC @== 1
+      cbFirstRPC @== 1
+      cbLastRPC @== 1
+
+      runInvariantsChecks
+
+    let incr = 5
+    -- Note: this does /not/ trigger new record because we are at the same time moment
+    call cfmm (Call @"Increase_observation_count") incr
+
+    consumer <- originateSimple "consumer" [] contractConsumer
+
+    storageSnapshot0 <- do
+      CumulativesBufferRPC{..} <- sCumulativesBufferRPC <$> getStorage cfmm
+      cbReservedLengthRPC @== 1 + incr
+      cbFirstRPC @== 1
+      cbLastRPC @== 1
+
+      runInvariantsChecks
+
+    -- We already registered one value, so performing one less probe
+    storageSnapshots1 <- for [1 .. incr - 1] \i -> do
+      -- Aribtrary call to put a new record in the buffer
```

Comment:
nitpick: `Aribtrary` -> `Arbitrary`

note: there's at least one more place with this typo